feat: Employee Grievance

Anurag810 opened the pull request:
**Documentation PR:** https://github.com/frappe/erpnext_documentation/pull/326

<h1>Employee Grievance</h1>

- [x] Development
- [x] Tests
- [x] Docs

### Grievance Type:
Users can define Grievance Type like. Pay and Benefits and Interpersonal Grievance like "Employee Abuse". And same features. (Refer screen Shot)
![image](https://user-images.githubusercontent.com/32095923/118123653-3108bf80-b412-11eb-82ec-4e21d912275a.png)
 
### Employee Grievance:
![image](https://user-images.githubusercontent.com/32095923/118124261-f9e6de00-b412-11eb-84e1-117ca282ef9b.png)

### Validation:
#### Employee:

1. You cannot change the status of an employee from suspended When an Employee is Suspended. (He will automatically Unsuspended after **"suspended to"** date. our you can unsuspend him manually)

#### Employee Grievance
1. Only Employee Grievance with status Invalid and Resolved can be submitted.
2. You need to Unsuspend Employee before cancellation.
3. You need to set the Cause of Grievance before setting Status to Investigated.
4. You need to fill Resolution Detail section before setting Employee Grievance Resolved.

### Workflow Flow:
Step 1:The employee will Create an Employee Grievance against the Employee or Group of Employees.
For process may require Some document to be linked
Select Grievance type saves the document.

Step 2:
The HR manager will investigate and he will fill the cause of grievance, resolution detail
The employee responsible for Grievance will have a pay cut or suspension(he will be suspended for suspension days). 
the status will change from **Active** to **Suspended** in employee master

Step3:
He can also unsuspend Employee.

Anurag810 committed:
feat: Employee Grievance

Anurag810 committed:
feat: Employee Grievance

Anurag810 committed:
feat: link to desk and automatic unsuspend

Anurag810 committed:
feat: link to desk and automatic unsuspend

Anurag810 committed:
test: Employee Grievance

Anurag810 committed:
test: Employee Grievance

Anurag810 committed:
fix: Sider and Translation

Anurag810 committed:
fix: Sider and Translation

Anurag810 committed:
fix: sider

Anurag810 committed:
fix: sider

Anurag810 committed:
fix: formatting

Anurag810 committed:
fix: formatting

Anurag810 committed:
Merge branch 'Employee-grievance' of https://github.com/anurag810/erpnext into Employee-grievance

ruchamahabal commented on erpnext/hr/doctype/grievance_type/grievance_type.json:
```diff
@@ -0,0 +1,97 @@
+   "default": "0",
+   "fieldname": "is_applicable_for_pay_cut",
+   "fieldtype": "Check",
+   "label": "Is Applicable for Pay-cut "
```
```suggestion
   "label": "Is Applicable for Pay Cut "
```
We can also shift this checkbox to the left instead of adding a column break since it is only 1 field.

Anurag810 replied:
Done

ruchamahabal commented on erpnext/hr/doctype/grievance_type/grievance_type.json:
```diff
@@ -0,0 +1,97 @@
+   "depends_on": "eval: doc.is_applicable_for_suspension == 1",
+   "fieldname": "number_of_days_for_suspension",
+   "fieldtype": "Int",
+   "label": "Number of days for Suspension",
```
```suggestion
   "label": "Suspension Period in Days",
```

Anurag810 replied:
Done

ruchamahabal commented on erpnext/hr/doctype/grievance_type/grievance_type.json:
```diff
@@ -0,0 +1,97 @@
+   "fieldtype": "Text",
+   "label": "Description"
+  }
+ ],
```
Add these fields to the List View and Standard Filter:

1. Is Applicable for Suspension
2. Is Applicable for Pay Cut

Anurag810 replied:
Dane

ruchamahabal commented on erpnext/hr/doctype/employee_grievance/employee_grievance.json:
```diff
@@ -0,0 +1,292 @@
+ "engine": "InnoDB",
+ "field_order": [
+  "raised_by",
+  "designation",
```
Let's also add `employee_name` to the fields.

Anurag810 replied:
added

ruchamahabal commented on erpnext/hr/doctype/employee_grievance/employee_grievance.json:
```diff
@@ -0,0 +1,292 @@
+  {
+   "fieldname": "description",
+   "fieldtype": "Text",
+   "label": "Description",
```
Instead of adding this label to the field, let's add a label for the section since there is only 1 field in this section.

Anurag810 replied:
![image](https://user-images.githubusercontent.com/32095923/119314743-b736c880-bc92-11eb-9ff6-960fd3d6d592.png)
As the field is mandatory * is not looking good. named field not section

ruchamahabal commented on erpnext/hr/doctype/employee_grievance/employee_grievance.json:
```diff
@@ -0,0 +1,292 @@
+   "fetch_from": "grievance_type.is_applicable_for_pay_cut",
+   "fieldname": "is_applicable_for_pay_cut",
+   "fieldtype": "Check",
+   "label": "Is Applicable for Pay-cut ",
```
```suggestion
   "label": "Is Applicable for Pay Cut ",
```

Anurag810 replied:
Done

ruchamahabal commented on erpnext/hr/doctype/employee_grievance/employee_grievance.json:
```diff
@@ -0,0 +1,292 @@
+   "fetch_from": "grievance_type.number_of_days_for_suspension",
+   "fieldname": "number_of_days_for_suspension",
+   "fieldtype": "Int",
+   "label": "Number of days for Suspension"
```
```suggestion
   "label": "Suspension Period in Days"
```

Anurag810 replied:
Done

ruchamahabal commented on erpnext/hr/doctype/employee_grievance/employee_grievance.json:
```diff
@@ -0,0 +1,292 @@
+   "fieldtype": "Select",
+   "in_list_view": 1,
+   "label": "Status",
+   "options": "Open\nInvestigated\nResolved\nInvalid",
```
```suggestion
   "options": "Open\nInvestigation\nResolved\nInvalid",
```
IMO, "Investigated" denotes a past action, so we should have the status as "Investigation" since this will track that grievance is under investigation.

Anurag810 replied:
https://github.com/frappe/erpnext/pull/25705#discussion_r635061636
As mentioned in this comment I think we should keep it Investigated as we need to track the root cause too...

ruchamahabal commented on erpnext/hr/doctype/employee_grievance/employee_grievance.py:
```diff
@@ -0,0 +1,112 @@
+		if self.status == "Investigated" and not self.cause_of_grievance:
+			frappe.throw(_("You ned to set Cause of Grievance before Setting Status to Investigated"))
+
+	def check_resolution_details(self):
+		raise_error = 0
+		mandatory_fields = []
+		if self.status == "Resolved":
+			if not self.resolution_date:
+				mandatory_fields.append("Resolution Date")
+				raise_error = 1
+
+			if not self.resolved_by:
+				mandatory_fields.append("Resolved By")
+				raise_error = 1
+
+			if not self.resolution_detail:
+				mandatory_fields.append("Resolution Detail")
+				raise_error = 1
+
+			if not self.cause_of_grievance:
+				mandatory_fields.append("Cause of Grievance")
+				raise_error = 1
+
+
+		fields_text = bold(comma_and(mandatory_fields))
+		if raise_error:
+			frappe.throw(_("You need to set {0} before setting Employee Grievance Resolved").format(fields_text))
```
Instead of adding validation for this, can't we set `mandatory_depends_on` property to make these fields mandatory on marking status as Resolved?

1. Resolved By
2. Resolution Date

Resolution Details should be optional. I don't see how the Cause of Grievance is any different than Resolution Details since we are already capturing grievance description and resolution details. If this field is to be retained, it should be a part of the Resolution Details section.

Anurag810 replied:
@ruchamahabal


- "**Description**" is actually what happened to the employee.  
- "**Cause of grievance**" is like, why this grievance happened in the first place, finding the root cause.  it is the result of the investigation
-  "**resolution detail**" are details for solution and prevention for the problem raised by the employee. so that, this grievance will not be happening again to employees.
-  "**action**' is a punishment given like Suspension and Pay Cut

I think these are the important matrices to be captured for complete grievance resolution.
for example, I filled the grievance form please refer to the screenshot in the description.

ruchamahabal replied:
Okay, then we should add the label "Investigation Details" for the Cause of Grievance section. So there will be a clear distinction:

1. Description
2. Investigation
3. Action
4. Resolution

Anurag810 replied:
Done

ruchamahabal commented on erpnext/hr/doctype/employee_grievance/employee_grievance.py:
```diff
@@ -0,0 +1,112 @@
+		if self.suspended_from and self.suspended_to:
+			frappe.throw(_("You need to Unsuspend Employee before Cancellation"))
+
+	def check_cause_of_grievance(self):
+		if self.status == "Investigated" and not self.cause_of_grievance:
+			frappe.throw(_("You ned to set Cause of Grievance before Setting Status to Investigated"))
```
As we are changing the status to "Investigation" rather than "Investigated", we should not have this validation.
Again I do not see a point in keeping so many fields for recording the incident:

Description
Cause of Grievance
Resolution Details

I guess we can remove "Cause of Grievance"

Anurag810 replied:
reference: https://github.com/frappe/erpnext/pull/25705#discussion_r635061636
As mentioned in this comment I think we should keep it Investigated as we need to track the root cause too...

ruchamahabal commented on erpnext/hr/doctype/employee_grievance/employee_grievance.js:
```diff
@@ -0,0 +1,114 @@
+				}
+			};
+		});
+		frm.set_query('associated_document_type', function() {
+			return {
+				filters: {
+					name: ['in', [
+						"Salary Slip",
+						"Payroll",
+						"Employee Benefits",
+						"Expense Claim",
+						"Leave Encashment",
+						"Employee Incentive"
```
I think we should keep these options open for selection except for child tables and single doctypes. Users can have custom doctypes or even other doctypes like Tasks, Projects, Issue, etc.

Anurag810 replied:
yeah we can make it generic but need to remove some of the core doctypes.

Anurag810 replied:
done

ruchamahabal commented on erpnext/hr/doctype/employee_grievance/employee_grievance.js:
```diff
@@ -0,0 +1,114 @@
+			}
+
+			if (frm.doc.is_applicable_for_pay_cut) {
+				frm.add_custom_button(__("Apply Pay-cut"), function () {
```
```suggestion
				frm.add_custom_button(__("Apply Pay Cut"), function () {
```

Anurag810 replied:
done

ruchamahabal commented on erpnext/hr/doctype/employee_grievance/employee_grievance.js:
```diff
@@ -0,0 +1,114 @@
+
+			if (frm.doc.is_applicable_for_pay_cut) {
+				frm.add_custom_button(__("Apply Pay-cut"), function () {
+					frm.events.create_additional_salary(frm, );
```
```suggestion
					frm.events.create_additional_salary(frm);
```

Anurag810 replied:
done

ruchamahabal commented on erpnext/hr/doctype/employee_grievance/employee_grievance.js:
```diff
@@ -0,0 +1,114 @@
+		}
+	},
+	suspend_or_unsuspend_employee: function(frm, action) {
+		let message = '';
+		let method = '';
+		if (action === "suspend") {
+			method = "erpnext.hr.doctype.employee_grievance.employee_grievance.suspend_employee";
+			message =  __('Are you sure you want to Suspend');
+			message += " "+frm.doc.employee_responsible+'?';
+		} else if (action === "unsuspend") {
+			method = "erpnext.hr.doctype.employee_grievance.employee_grievance.unsuspend_employee";
+			message = __('Are you sure you want to Un-suspend');
+			message += " "+frm.doc.employee_responsible+'?';
+		}
+
+		if (frm.doc.employee_responsible) {
+			frappe.msgprint({
+				title: __('Notification'),
+				message: message,
+				primary_action: {
+					label: __('Yes'),
+					action() {
+						frappe.call({
+							method: method,
+							args: {
+								name: frm.doc.name
+							},
+							callback: function () {
+								frm.refresh();
+								cur_dialog.hide();
+							}
+						});
+					}
+				}
+			});
+		}
```
1. Use `frappe.confirm` instead of the following dialog.
2. Also, the string formation is not very translation-friendly.
3. We should also try and commonify the server-side functions.
Check refactored code below:

![image](https://user-images.githubusercontent.com/24353136/118648010-9805ea00-b7ff-11eb-9d55-505f3793590e.png)

```suggestion	
		let message = '';
		let method = '';

		if (action === 'suspend') {
			method = 'erpnext.hr.doctype.employee_grievance.employee_grievance.suspend_employee';
			message =  __('Are you sure you want to suspend the employee {0}', [frm.doc.employee_responsible]);
		} else if (action === 'unsuspend') {
			method = 'erpnext.hr.doctype.employee_grievance.employee_grievance.unsuspend_employee';
			message = __('Are you sure you want to unsuspend the employee {0}', [frm.doc.employee_responsible]);
		}

		if (frm.doc.employee_responsible) {
			frappe.confirm((message),
				function() {
					frappe.call({
						method: method,
						args: {
							name: frm.doc.name
						},
						callback: function() {
							frm.refresh();
							window.close();
						}
					});
				},
				function() {
					window.close();
				}
			);
		}
```

Anurag810 replied:
Done

ruchamahabal commented on erpnext/hr/doctype/employee_grievance/employee_grievance.js:
```diff
@@ -0,0 +1,114 @@
+			}
+		});
+	},
+
```
Try and keep uniform quote styles throughout a single file.

ruchamahabal commented on erpnext/hr/doctype/employee_grievance/employee_grievance.py:
```diff
@@ -0,0 +1,112 @@
+		additional_salary.to_date = doc.suspended_to
+		additional_salary.type = "Deduction"
+	else:
+		frappe.throw(_("Paycut is already created"))
```
```suggestion
		frappe.throw(_("Additional Salary of type Deduction is already created for pay cut"))
```

Anurag810 replied:
done

ruchamahabal commented on erpnext/hr/doctype/employee_grievance/employee_grievance.py:
```diff
@@ -0,0 +1,112 @@
+	return additional_salary
+
+
+@frappe.whitelist()
+def unsuspend_employee(name):
+	frappe.db.set_value("Employee Grievance", name, "suspended_from", None)
+	frappe.db.set_value("Employee Grievance", name, "suspended_to", None)
+	frappe.db.set_value("Employee Grievance", name, "unsuspended_on", today())
+	employee_responsible = frappe.db.get_value("Employee Grievance", name, "employee_responsible")
+	frappe.db.set_value("Employee", employee_responsible, "status", "Active")
+
+
+@frappe.whitelist()
+def suspend_employee(name):
+	employee_grievance = frappe.get_doc("Employee Grievance", name)
+	today_date = get_datetime(today())
+	days = employee_grievance.number_of_days_for_suspension
+
+	if not days:
+		frappe.throw(_("Grievance Type: {0} is not applicable for suspension").format(employee_grievance.grievance_type))
+
+	employee_grievance.db_set('suspended_from', today())
+	employee_grievance.db_set('suspended_to', add_days(today_date, days))
+
+	frappe.db.set_value("Employee", employee_grievance.employee_responsible, "status", "Suspended")
```
1. IMO, these functions should be commonified.
2. Also, we should not reset "suspended from" and "suspended to" on unsuspension. The grievance record should maintain the suspension dates even after the suspension period is over.
3. When setting values in db, you can do it in 1 call:

```python
frappe.db.set_value("Employee Grievance", name, {
	"suspended_from": None,
	"suspended_to": None
})
```

Anurag810 replied:
done

ruchamahabal commented on erpnext/hr/doctype/employee_grievance/employee_grievance.py:
```diff
@@ -0,0 +1,112 @@
+		"docstatus": 1,
+		"suspended_to": ('<', today()),
+	}, fields = ["name", "suspended_to"])
+	print(data)
```
```suggestion
```

Anurag810 replied:
done

ruchamahabal commented on erpnext/hr/doctype/employee_grievance/employee_grievance.py:
```diff
@@ -0,0 +1,112 @@
+	frappe.db.set_value("Employee", employee_grievance.employee_responsible, "status", "Suspended")
+
+@frappe.whitelist()
+def check_and_unsuspend_employees():
```
Again, this function has to be changed as we should retain "suspended from" and "suspended to". Add some flag, like `employee_suspended`, and filter records using it.

ruchamahabal commented on erpnext/hr/doctype/employee/employee.py:
```diff
@@ -173,6 +173,26 @@ def validate_date(self):
 			throw(_("Contract End Date must be greater than Date of Joining"))
 
+	def validated_suspension(self):
+		if self.status != "Suspended":
```
You can reduce the number of times this runs by checking `doc_before_save`
```suggestion
		doc_before_save = self.get_doc_before_save()
		if self.status != "Suspended" and doc_before_save.status == "Suspended":
```

Anurag810 replied:
done

ruchamahabal commented on erpnext/hr/doctype/employee/employee.py:
```diff
@@ -173,6 +173,26 @@ def validate_date(self):
+			if len(grievance):
+				self.db_set("status", "Suspended")
+				grievance = grievance[0]
+				frappe.throw(_("You are not allowed to change status, due to suspension from {0} to {1}. Reference: {2}").format(
+					bold(format_date(grievance.suspended_from, )),
+					bold(format_date(grievance.suspended_to)),
+					bold(get_link_to_form("Employee Grievance", grievance.name))
+				))
```
```suggestion
				frappe.throw(_("You are not allowed to change the status. The employee has been suspended from {0} to {1} due to Employee Grievance {2}").format(
					bold(format_date(grievance.suspended_from)),
					bold(format_date(grievance.suspended_to)),
					bold(get_link_to_form("Employee Grievance", grievance.name))
				), title=_("Not Allowed"))
```

Anurag810 replied:
done

ruchamahabal commented on erpnext/hr/doctype/employee_grievance/employee_grievance.js:
```diff
@@ -0,0 +1,114 @@
+				let message_line1 = "Employee: <b>"+ frm.doc.employee_responsible +"</b> is suspended from <b>"+ suspended_from+"</b> to <b>" +suspended_to + "</b>.";
+				let message_line2 = "Employee will be unsuspended automatically or you can do it manually by clicking on unsuspend Employee.";
+
+				let html = '<span class="indicator whitespace-nowrap orange"><span>';
```
Keep the indicator only if the message line does not span 2 lines as the alignment looks off.

![image](https://user-images.githubusercontent.com/24353136/118765965-6e999c80-b899-11eb-89fc-2e6e3779ab8c.png)

ruchamahabal commented on erpnext/hr/doctype/employee_grievance/employee_grievance.js:
```diff
@@ -0,0 +1,114 @@
+				let html = '<span class="indicator whitespace-nowrap orange"><span>';
+				html += message_line1;
+				html += '</span></span><br><span>';
+				html += message_line2+'</span>';
```
You can construct the entire message using backticks ``. It's easier to read, maintain, and debug.

Check reference: https://github.com/frappe/erpnext/blob/d15d0d8568cfff8f0477374572b73a17fe44fb0e/erpnext/crm/doctype/linkedin_settings/linkedin_settings.js#L20-L27

Anurag810 replied:
removed and kept simple

ruchamahabal commented on erpnext/hr/doctype/employee_grievance/employee_grievance.js:
```diff
@@ -0,0 +1,114 @@
+				let suspended_from = frappe.datetime.global_date_format(frm.doc.suspended_from);
+				let suspended_to = frappe.datetime.global_date_format(frm.doc.suspended_to);
+
+				let message_line1 = "Employee: <b>"+ frm.doc.employee_responsible +"</b> is suspended from <b>"+ suspended_from+"</b> to <b>" +suspended_to + "</b>.";
+				let message_line2 = "Employee will be unsuspended automatically or you can do it manually by clicking on unsuspend Employee.";
```
Not translation-friendly. This message is too long.

```suggestion
				let message = __("Employee {0} is suspended from {1} to {2}. {0} will be unsuspended automatically on {2}.", [frm.doc.employee_responsible, suspended_from, suspended_to]);
```
The button label is clear enough to indicate that the user can manually unsuspend the employee. Do we need to mention it explicitly?

Anurag810 replied:
done

ruchamahabal commented on erpnext/hr/doctype/employee_grievance/employee_grievance.py:
```diff
@@ -0,0 +1,112 @@
+
+	def before_cancel(self):
+		if self.suspended_from and self.suspended_to:
+			frappe.throw(_("You need to Unsuspend Employee before Cancellation"))
```
Additional salary record also needs to be canceled on the `on_cancel` event

Anurag810 replied:
No need to as it, already give a message for the linked document during cancelation.

ruchamahabal commented on erpnext/hr/doctype/employee_grievance/employee_grievance.json:
```diff
@@ -0,0 +1,292 @@
+  {
+   "fieldname": "suspended_from",
+   "fieldtype": "Date",
+   "label": "Suspended from ",
```
```suggestion
   "label": "Suspended From",
```

Anurag810 replied:
done

ruchamahabal commented on erpnext/hr/doctype/employee_grievance/employee_grievance.json:
```diff
@@ -0,0 +1,292 @@
+  {
+   "fieldname": "suspended_to",
+   "fieldtype": "Date",
+   "label": "Suspended to",
```
```suggestion
   "label": "Suspended To",
```

Anurag810 replied:
done

ruchamahabal commented on erpnext/hr/doctype/employee_grievance/employee_grievance.json:
```diff
@@ -0,0 +1,292 @@
+  {
+   "fieldname": "unsuspended_on",
+   "fieldtype": "Date",
+   "label": "Unsuspended on",
```
```suggestion
   "label": "Unsuspended On",
```

Anurag810 replied:
done

ruchamahabal commented on erpnext/hr/doctype/employee_grievance/employee_grievance.json:
```diff
@@ -0,0 +1,292 @@
+   "fetch_from": "raised_by.reports_to",
+   "fieldname": "reports_to",
+   "fieldtype": "Link",
+   "label": "Reports to",
```
```suggestion
   "label": "Reports To",
```

Anurag810 replied:
done

ruchamahabal commented on erpnext/hr/doctype/employee_grievance/employee_grievance.json:
```diff
@@ -0,0 +1,292 @@
+   "fieldname": "grievance_against_party",
+   "fieldtype": "Link",
+   "in_list_view": 1,
+   "label": "Grievance Against Party ",
```
```suggestion
   "label": "Grievance Against Party",
```

Anurag810 replied:
done

ruchamahabal commented on erpnext/controllers/queries.py:
```diff
@@ -18,7 +18,7 @@ def employee_query(doctype, txt, searchfield, start, page_len, filters):
 
 	return frappe.db.sql("""select {fields} from `tabEmployee`
-		where status = 'Active'
+		where status in ('Active', 'Suspended')
```
I don't think we should allow selecting Suspended employees.

Anurag810 replied:
Allowed because they may be part of payroll process and several other features... like suspended employee can be the part of payroll and can have their salary slip. They may also have leave policy assignments, they may also appear in Leave allocation and lots more process....

Anurag810 committed:
feat: changes requested

Anurag810 committed:
feat: changes requested

Anurag810 commented:
@ruchamahabal done changes

ruchamahabal committed:
fix: patch test and sider issue

ruchamahabal committed:
fix: make Employee Responsible non-mandatory since there cannot be an employee responsible for all sorts of grievances

- show pay cut and suspension buttons only if Employee Resposible is set

- some label changes

ruchamahabal committed:
feat: added subject field for more context

- set title for documents

- added list view settings

- refactor suspend and unsuspend functions

- add submit and cancel perms for system and hr managers

- fix tests

ruchamahabal committed:
fix: sider issues

ruchamahabal commented on erpnext/hr/doctype/employee_grievance/employee_grievance.py:
```diff
@@ -0,0 +1,86 @@
+
+
+@frappe.whitelist()
+def check_and_unsuspend_employees():
```
@Anurag810 also add a test case for this hook function.

rmehta commented on erpnext/hr/doctype/employee_grievance/employee_grievance.js:
```diff
@@ -0,0 +1,107 @@
+		});
+	},
+
+	refresh: function(frm) {
```
unnecessary client side code

rmehta replied:
this should be implemented via workflow

Anurag810 replied:
These are good to have features when dealing with grievances. It's just code to add custom buttons for additional salary creation or changing employee status to Suspended. Employees can be got suspended for Employee Grievance like Employee Abuse, bullying. 

I am not sure how workflow doctype can handle document action mapping?
It just allows setting status or updating some value in the same document

yadavyk replied:
Based on the selection of employee made in the "Raised By" field, the selected employee name should be refrained from getting reflected in the dropdown of "Grievance Against" field. 

Screenshot for reference: 

![Employee Grievance ](https://user-images.githubusercontent.com/32797974/122364940-debe3f80-cf77-11eb-9c3b-9176f91d3e94.png)

Anurag810 replied:
fixed

Anurag810 committed:
fix: removed suspension and paycut

Anurag810 committed:
fix:sider

Anurag810 committed:
Merge branch 'Employee-grievance' of https://github.com/anurag810/erpnext into Employee-grievance

Anurag810 committed:
fix: test

Anurag810 committed:
fix: test

Anurag810 committed:
fix: resolved Conflicts

Anurag810 committed:
fix: sider

ruchamahabal commented on erpnext/payroll/doctype/salary_structure/test_salary_structure.py:
```diff
@@ -65,6 +65,7 @@ def test_amount_totals(self):
 		frappe.db.set_value("Payroll Settings", None, "include_holidays_in_total_working_days", 0)
 		sal_slip = frappe.get_value("Salary Slip", {"employee_name":"test_employee_2@salary.com"})
 		if not sal_slip:
+			print("I am Here")
```
```suggestion
```

ruchamahabal commented on erpnext/payroll/doctype/salary_slip/test_salary_slip.py:
```diff
@@ -478,11 +478,14 @@ def get_no_of_days(self):
 def make_employee_salary_slip(user, payroll_frequency, salary_structure=None):
 	from erpnext.payroll.doctype.salary_structure.test_salary_structure import make_salary_structure
 
+	print(user, payroll_frequency, salary_structure)
```
```suggestion
```

ruchamahabal commented on erpnext/payroll/doctype/salary_slip/test_salary_slip.py:
```diff
@@ -478,11 +478,14 @@ def get_no_of_days(self):
+
 	employee = frappe.db.get_value("Employee", {"user_id": user})
 	salary_structure_doc = make_salary_structure(salary_structure, payroll_frequency, employee=employee)
+	print(salary_structure_doc)
```
```suggestion
```

ruchamahabal committed:
fix: remove debugging print statements

ruchamahabal commented on erpnext/hr/doctype/employee_grievance/employee_grievance.py:
```diff
@@ -0,0 +1,16 @@
+class EmployeeGrievance(Document):
+	def on_submit(self):
+		if self.status not in ["Invalid", "Resolved"]:
+			frappe.throw(_("Only Employee Grievance with status {0} and {1} can be submitted").format(
```
```suggestion
			frappe.throw(_("Only Employee Grievance with status {0} or {1} can be submitted").format(
```

ruchamahabal committed:
fix: validation message

ruchamahabal commented on erpnext/hooks.py:
```diff
@@ -363,7 +363,7 @@
 		"erpnext.buying.doctype.supplier_quotation.supplier_quotation.set_expired_status",
 		"erpnext.accounts.doctype.process_statement_of_accounts.process_statement_of_accounts.send_auto_email",
-		"erpnext.non_profit.doctype.membership.membership.set_expired_status"
+		"erpnext.non_profit.doctype.membership.membership.set_expired_status",
```
```suggestion
		"erpnext.non_profit.doctype.membership.membership.set_expired_status"
```

ruchamahabal committed:
fix: unnecessary comma

Anurag810 commented:
Pay-cut and  Employee suspension is a good feature to have, Will cover that part in different PR. with different workflow and design.

ruchamahabal commented:
@Anurag810 Please update the description and docs with revised screenshots
